proposals/infra/INF-0004-register-types-and-diagnostics.md

```diff
@@ -0,0 +1,250 @@
+are either `basic`, `resource`, `udt`, or `other`. `basic` refers to a numeric variable
+type. `other` refers to a type that cannot possibly be resource, because it lacks the 
+resource attribute.<br>
+The next group of flags are the resource class flags,
+which are `srv`, `uav`, `cbv`, or `sampler`.<br>
+The final flag is `default_globals`, which indicates whether or not the decl appears
+inside the $Globals scope.<br><br>
+
+From the Decl object, first determine if the decl is a cbuffer or tbuffer decl. If so,
+we know the decl is a `resource` and can infer the resource class (`cbv` or `srv`
+respectively). Otherwise, the decl is classified according to the underlying canonical 
+type.<br>
+If the decl is implicit, then check if the decl is a vector, matrix, or otherwise numeric type,
+and if so, we know that the decl is a `basic` type. If the decl has a resource attribute, 
+we can set the `resource` flag, and can also obtain the resource class and set the resource 
+class flag accordingly. Otherwise, we know the decl is not a resource and not basic, and so
+the `other` flag is set.<br>
+If the decl is not implicit, then we check if it is a struct or class type, and if so, the
+decl is a UDT, so the `udt` flag is set. We then recurse through the members of the UDT and
+collect information on what types of resources are contained by the UDT (uav, cbv, srv, or
+sampler). The corresponding resouce class flags are set. If the decl is not a struct or class
+type, then it is a basic type, and so the `basic` flag is set.<br>
+The last step is to simply check if the decl is in the `$Globals` scope. If so, then the
+`default_globals` flag will be set.<br><br>
+
+Now enough information has been gathered to determine the right diagnostics to emit, if any.
+Firstly, if `other` is set, then the variable type is not valid for the given register type.
+So, we can emit this diagnostic: `error: '%0' is an invalid resource type for register type '%1'`
+If the `resource` flag is set and the `udt` flag is not set, then we have to check the 
+register statement, and the resource class flag, and verify that the register type matches the 
+resource class flag. If not, we emit this diagnostic: 
+`error: %select{SRV|UAV|CBV|Sampler}2 type '%0' requires register type '%select{t|u|b|s}2', but register type '%1' was used.`
+If the `basic` flag is set, then we first check if `default_globals` is set. If so, then 
+we check the register type. If the register type is 'i' or 'b', the deprecated warning
+that is treated as an error as shown in the examples above will be emitted. If 'c' is given,
+no errors will be emitted. If `default_globals` is not set and 'c' is given, a warning will
+be emitted that is treated as an error by default: `warning: register binding 'c' should only be used in global contexts. Disable with --Wno-disallow-legacy-binding-rules` 
+After this point, `default_globals` doesn't need to be set.

```

Comment:
Suggested rewording, and `-Wno-disallow-legacy-binding-rules` shouldn't be part of the message, should it?  It will emit that warning group as part of the message automatically.  Also, I think `pack_offset` should be suggested, since it does the thing that would otherwise be done by the `c` binding inside cbuffer/tbuffer decls.

```suggestion
that is treated as an error as shown in the examples above will be emitted.
If the register type is 'c', it is allowed, but if `default_globals` is not set,
a warning that is treated as an error by default will be emitted:
`warning: register binding 'c' ignored inside cbuffer/tbuffer declarations; use pack_offset instead`.
This warning is part of the warning group `disallow-legacy-binding-rules`.

```